src/blog/understanding-ecmascript-part-3.md

```diff
@@ -0,0 +1,320 @@
+
+### Lexical grammar
+
+The [lexical grammar](https://tc39.es/ecma262/#sec-ecmascript-language-lexical-grammar) describes how Unicode code points are translated into a sequence of **input elements** (tokens, line terminators, comments, white space).
+
+There are several cases where the next token cannot be identified purely by looking at the Unicode code point stream, but we need to know where we are in the syntactic grammar. A classic example is `/`. To know whether it's a division or the start of the RegExp, we need to know which one is allowed in the syntactic context we're currently in.
+
+For example:
+```javascript
+const x = 10 / 5;
+//           ^ this is a DivPunctuator
+
+const r = /foo/;
+//        ^ this is the start of a RegularExpressionLiteral
+```
+
+A similar thing happens with templates &mdash; the interpretation of <code>{`</code> depends on the context we're in:
+
+```javascript
+const what1 = 'temp';
+const what2 = 'late';
+const t = `I am a ${ what1 + what2 }`;
+// `I am a ${ is TemplateHead
+// }` is TemplateTail
+
+if (0 == 1) {
+}`not very useful`;
+// } is RightBracePunctuator
+// ` is the start of a NoSubstitutionTemplate
+
+```
+
+The lexical grammar uses several goal symbols to distinguish between the contexts where some input elements are permitted and some are not. For example, the goal symbol `InputElementDiv` is used in contexts where `/` is a division and `/=` is a division-assignment. The `InputElementDiv` productions list the possible tokens which can be produced in this context:
+
+> [`InputElementDiv ::`](https://tc39.es/ecma262/#prod-InputElementDiv)
+> `WhiteSpace`
+> `LineTerminator`
+> `Comment`
+> `CommonToken`
+> `DivPunctuator`
+> `RightBracePunctuator`
+
+In this context, encountering `/` will produce the `DivPunctuator` input element. Producing a `RegularExpressionLiteral` is not an option here.
+
+On the other hand, `InputElementRegExp` is the goal symbol for the contexts where `/` is the beginning of a RegExp:
+
+> [`InputElementRegExp ::`](https://tc39.es/ecma262/#prod-InputElementRegExp)
+> `WhiteSpace`
+> `LineTerminator`
+> `Comment`
+> `CommonToken`
+> `RightBracePunctuator`
+> `RegularExpressionLiteral`
+
+As we see from the productions, it's possible that this produces the `RegularExpressionLiteral` input element, but producing `DivPunctuator` is not possible.
+
+Similarly, there is another goal symbol, `InputElementRegExpOrTemplateTail`, for contexts where `TemplateMiddle` and `TemplateTail` are permitted, in addition to `RegularExpressionLiteral`. And finally, `InputElementTemplateTail` is the goal symbol for contexts where only `TemplateMiddle` and `TemplateTail` are permitted but `RegularExpressionLiteral` is not permitted.
+
+We can imagine the syntactic grammar analyzer ("parser") calling the lexical grammar analyzer ("tokenizer" or "lexer"), passing the goal symbol as a parameter and asking for the next input element suitable for that goal symbol.
+
+### Other grammars
```

Comment:
`Other grammars` seems like a weird header. However, I think you can fix this by introducing the "table of contents" I proposed above.

Reply:
I moved them into the introduction chapter at the top. The long explanations of the lexical & syntactic grammars comes after that.